Predict whether a student will Pass or Fail based on:

Study Hours
Attendance
| Student | Study Hours | Attendance | Result |
| ------- | ----------- | ---------- | ------ |
| 1       | High        | High       | Pass   |
| 2       | High        | Low        | Pass   |
| 3       | Low         | High       | Pass   |
| 4       | Low         | Low        | Fail   |
| 5       | High        | High       | Pass   |
| 6       | Low         | Low        | Fail   |
| 7       | High        | Low        | Pass   |
| 8       | Low         | High       | Fail   |

We want to predict:

Study Hours = High and Attendance = High

Will the student Pass or Fail?


The Naïve Assumption

The word Naïve refers to the assumption that all features are conditionally independent given the class.

This means

Study Hours and Attendance are assumed to be independent once we know whether the student passes or fails.

P(Study,Attendance∣Pass)=P(Study∣Pass)×P(Attendance∣Pass)



(P(A)) = Prior Probability
•	(P(B|A)) = Likelihood
•	(P(A|B)) = Posterior Probability
•	(P(B)) = Evidence

P(A/B)=P(B/A)P(A)/P(B)

P(Class/Features)=P(Features/Class)P(CLass)/P(Features)

In [2]:
# ======================================================
# Naive Bayes Classifier from Scratch
# ======================================================

import pandas as pd

# -----------------------------------------------------
# Step 1 : Create Dataset
# -----------------------------------------------------

data = {
    'Study':['High','High','Low','Low',
             'High','Low','High','Low'],

    'Attendance':['High','Low','High','Low',
                  'High','Low','Low','High'],

    'Result':['Pass','Pass','Pass','Fail',
              'Pass','Fail','Pass','Fail']
}

df = pd.DataFrame(data)

print("Training Dataset\n")
print(df)

# -----------------------------------------------------
# Step 2 : Prior Probabilities
# -----------------------------------------------------

total = len(df)

pass_count = len(df[df['Result']=="Pass"])
fail_count = len(df[df['Result']=="Fail"])

P_pass = pass_count/total
P_fail = fail_count/total

print("\nPrior Probabilities")

print("P(Pass) =",P_pass)
print("P(Fail) =",P_fail)

# -----------------------------------------------------
# Step 3 : Likelihood Probabilities
# -----------------------------------------------------

P_study_high_pass = len(df[(df['Study']=="High") &
                           (df['Result']=="Pass")])/pass_count

P_att_high_pass = len(df[(df['Attendance']=="High") &
                         (df['Result']=="Pass")])/pass_count

P_study_high_fail = len(df[(df['Study']=="High") &
                           (df['Result']=="Fail")])/fail_count

P_att_high_fail = len(df[(df['Attendance']=="High") &
                         (df['Result']=="Fail")])/fail_count

print("\nLikelihoods")

print("P(Study=High | Pass) =",P_study_high_pass)
print("P(Attendance=High | Pass) =",P_att_high_pass)

print("P(Study=High | Fail) =",P_study_high_fail)
print("P(Attendance=High | Fail) =",P_att_high_fail)

# -----------------------------------------------------
# Step 4 : Naive Bayes Calculation
# -----------------------------------------------------

pass_score = P_pass * P_study_high_pass * P_att_high_pass

fail_score = P_fail * P_study_high_fail * P_att_high_fail

print("\nPosterior Scores")

print("Pass Score =",pass_score)
print("Fail Score =",fail_score)

# -----------------------------------------------------
# Step 5 : Prediction
# -----------------------------------------------------

if pass_score > fail_score:
    prediction="PASS"
else:
    prediction="FAIL"

print("\nPrediction =",prediction)

Training Dataset

  Study Attendance Result
0  High       High   Pass
1  High        Low   Pass
2   Low       High   Pass
3   Low        Low   Fail
4  High       High   Pass
5   Low        Low   Fail
6  High        Low   Pass
7   Low       High   Fail

Prior Probabilities
P(Pass) = 0.625
P(Fail) = 0.375

Likelihoods
P(Study=High | Pass) = 0.8
P(Attendance=High | Pass) = 0.6
P(Study=High | Fail) = 0.0
P(Attendance=High | Fail) = 0.3333333333333333

Posterior Scores
Pass Score = 0.3
Fail Score = 0.0

Prediction = PASS


The prior probability represents the probability of each class before considering the feature values.

Count the number of students in each class.
Total Students = 8

Pass = 5

Fail = 3

P(Pass)=5/8=0.625

P(Fail)=3/8=0.375

In [4]:
# -----------------------------------------------------
# Step 2 : Prior Probabilities
# -----------------------------------------------------

total = len(df)

pass_count = len(df[df['Result']=="Pass"])
fail_count = len(df[df['Result']=="Fail"])

P_pass = pass_count/total
P_fail = fail_count/total

print("\nPrior Probabilities")

print("P(Pass) =",P_pass)
print("P(Fail) =",P_fail)


Prior Probabilities
P(Pass) = 0.625
P(Fail) = 0.375



Likelihood Probability

Next, calculate the probability of observing each feature value within each class.

Among the Pass students:
High Study = 4

Total Pass =5

P(Study=High | Pass)=4/5=0.8

SImilarly
High Attendance =3

P(Attendance=High | Pass)=3/5=0.6

In [5]:
# -----------------------------------------------------
# Step 3 : Likelihood Probabilities
# -----------------------------------------------------

P_study_high_pass = len(df[(df['Study']=="High") &
                           (df['Result']=="Pass")])/pass_count

P_att_high_pass = len(df[(df['Attendance']=="High") &
                         (df['Result']=="Pass")])/pass_count

P_study_high_fail = len(df[(df['Study']=="High") &
                           (df['Result']=="Fail")])/fail_count

P_att_high_fail = len(df[(df['Attendance']=="High") &
                         (df['Result']=="Fail")])/fail_count

print("\nLikelihoods")

print("P(Study=High | Pass) =",P_study_high_pass)
print("P(Attendance=High | Pass) =",P_att_high_pass)

print("P(Study=High | Fail) =",P_study_high_fail)
print("P(Attendance=High | Fail) =",P_att_high_fail)


Likelihoods
P(Study=High | Pass) = 0.8
P(Attendance=High | Pass) = 0.6
P(Study=High | Fail) = 0.0
P(Attendance=High | Fail) = 0.3333333333333333


For the test case
Study = High

Attendance = High

COmpute

Pass Score= P(Pass) × P(Study=High | Pass) × P(Attendance=High | Pass)

Fail Score =


In [6]:
# -----------------------------------------------------
# Step 4 : Naive Bayes Calculation
# -----------------------------------------------------

pass_score = P_pass * P_study_high_pass * P_att_high_pass

fail_score = P_fail * P_study_high_fail * P_att_high_fail

print("\nPosterior Scores")

print("Pass Score =",pass_score)
print("Fail Score =",fail_score)


Posterior Scores
Pass Score = 0.3
Fail Score = 0.0
